### PR TITLE
Add EIP: Redeemable Tokens

### PR DESCRIPTION
**Pull request by LucazFFz**
A proposal for redeemable tokens.

### CODE REVIEW

**Comment by abcoathup** [comment]
EIP editors assign an EIP number (generally the PR number, but the decision is with the editors) (from: [EIP1](https://eips.ethereum.org/EIPS/eip-1#eip-numbers))

I couldn't see where an EIP editor assigned ERC4365.

**Comment by LucazFFz** [comment]
>

**Comment by abcoathup** [comment]
> @abcoathup Oh okay, I must have missed that. I apologize. I will remove the EIP number. However, it would be easier to keep this number since it's not taken and used extensively in the reference implementation code. I can absolutely change that if necessary.

Need to wait for an EIP editor to assign a number and then you can update the reference implementation.  I've had this before for an implementation, search and replace should resolve it quickly.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Please fix the CI errors.

We're experimenting with a new automated system to assign EIP numbers. Feel free to ignore the eip walidator error about the missing `eip` header for now, and avoid self-references. Simply refer to the current eip as "this EIP", rather than by "EIP-X".

For the code, simply use descriptive names for the classes. It's much more readable than an opaque number, anyway.

File: EIPS/eip-draft_title_abbrev.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-03-20
+requires: 165
+---
+
+## Abstract
+
+[ERC-4365](./eip-draft_title_abbrev.md) outlines a smart contract interface that can represent any number of fungible and non-fungible redeemable token types. The standard builds upon the [ERC-1155](./eip-1155.md) standard borrowing many of the ideas introduced by it including support for multiple tokens within the same contract and batch operations.
```
Comment (Pandapip1) [inline EIPS/eip-draft_title_abbrev.md:15]:
```suggestion
This EIP outlines a smart contract interface that can represent any number of fungible and non-fungible redeemable token types. The standard builds upon the [ERC-1155](./eip-1155.md) standard borrowing many of the ideas introduced by it including support for multiple tokens within the same contract and batch operations.
```

File: EIPS/eip-draft_title_abbrev.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+- **VIP tickets** - All-access tickets that include everything the event offers with activities excluded from the general admission tickets to give the attendees a premium experience. These tickets can be non-fungible making every ticket unique and allowing each ticket to be connected to an individual artwork or likewise to incentivize purchasing VIP tickets instead.
+
+Additionally, these tokens can be used to create, among others, vouchers, gift cards, and scratchcards.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+**Every smart contract implementing this ERC MUST implement the all of the functions in the `ERC4365` interface. Additionally, they MUST implement the ERC-165 `supportsInterface` function and MUST return the constant value `true` if the constant value `0x9d1da9d1` is passed through the `interfaceId` argument.**
```
Comment (Pandapip1) [inline EIPS/eip-draft_title_abbrev.md:34]:
```suggestion
**Every smart contract implementing this ERC MUST implement the all of the functions in the `IRedeemableToken` interface. Additionally, they MUST implement the [ERC-165](./eip-165.md) `supportsInterface` function and MUST return the constant value `true` if the constant value `0x9d1da9d1` is passed through the `interfaceId` argument.**
```

File: EIPS/eip-draft_title_abbrev.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+```solidity
+interface IERC4365 is IERC165 {
```
Comment (Pandapip1) [inline EIPS/eip-draft_title_abbrev.md:37]:
```suggestion
interface IRedeemableToken is IERC165 {
```

File: EIPS/eip-draft_title_abbrev.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+In addition, in order for a contract to be compliant with this ERC, it MUST abide by the following:
+
+- Implementers MAY enable token transfers to support functionality such as ticket reselling.
+- Implementers MUST NOT allow tokens to be transferred between addresses after they have been redeemed.
+- Implementers MUST allow token holders to redeem their tokens.
+- Implementers MUST NOT allow token issuers to redeem the tokens they have issued.
+- Implementers SHOULD allow token recipients to burn any token they receive.
+- Implementers MAY enable token issuers to burn the tokens they issued.
+
+**Smart contracts MUST implement all of the functions in the `ERC4365Receiver` interface to accept tokens being minted to them.**
```
Comment (Pandapip1) [inline EIPS/eip-draft_title_abbrev.md:134]:
```suggestion
**Smart contracts MUST implement all of the functions in the `IRedeemableTokenReceiver` interface to accept tokens being minted to them.**
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
You get the drill.

File: EIPS/eip-draft_title_abbrev.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
+---
```
[the remark this answers is not in the document]
Comment (abcoathup) [reply EIPS/eip-draft_title_abbrev.md:1]: I've assigned it as 6732, so you can update the EIP and your code.

**Review by mspaansen — APPROVED** [review]

**Comment by Pandapip1** [comment]
@eth-bot rerun

**Review by SamWilsn — COMMENTED** [review]
In addition to my other comments, would you mind reducing your reference implementation down to something more minimal? At minimum you should remove any of the OZ source. The readers can find those implementations themselves.

Generally the reference implementation should be one or two files to illustrate your proposal, and not a fully working production contract.

File: EIPS/eip-6732.md [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+category: ERC
+created: 2023-03-20
+requires: 165
+---
+
+## Abstract
+
+This ERC outlines a smart contract interface that can represent any number of fungible and non-fungible redeemable token types. The standard builds upon the [ERC-1155](./eip-1155.md) standard borrowing many of the ideas introduced by it including support for multiple tokens within the same contract and batch operations.
+
+Contrary to the ERC-1155 standard, this ERC does not enforce transferability as it recognizes situations where implementers might not want to allow it. Additionally, it introduces several extensions used to expand the functionality like the **Expirable extension** which provides a simple way to add an expiry date to tokens.
```
[the remark this answers is not in the document]
Comment (LucazFFz) [reply EIPS/eip-6732.md:18]: Yes that is probably a good idea

File: EIPS/eip-6732.md [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+This ERC takes inspiration from the ERC-1155 standard and only supports safe-style transfers when minting tokens. This enables receiver contracts to depend on `RedeemableTokenMint` or `RedeemableTokenMintBatch` functions to be called at the end of minting. This allows receivers to reject an increase in balance.
+
+## Backwards Compatibility
+
+Because of the heavy inspiration from the ERC-1155 standard many concepts and methods remain identical, namely the concept of a batch, the `balanceOf` and `balanceOfBatch` functions, and to some extent the URI Storage and Supply extensions.
+
+## Reference Implementation
+
+- [Reference implementation](../assets/eip-6732/)
```
[the remark this answers is not in the document]
Comment (LucazFFz) [reply EIPS/eip-6732.md:288]: Do you mean the reference implementation's readme?

File: assets/eip-6732/contracts/Migrations.sol
[the remark this answers is not in the document]
Comment (LucazFFz) [reply assets/eip-6732/contracts/Migrations.sol:1]: Would it be sufficient to keep the files under the examples and token folders and remove the rest including utils and migrations?

File: EIPS/eip-6732.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (SamWilsn) [reply EIPS/eip-6732.md:290]: Yeah, if you cram the whole reference implementation into a single file, just link to that. If not, you can make a `README.md` that's just a manual listing of files:

```markdown
# ERC-XXXX Reference Implementation

 * [SomeContract.sol](./SomeContract.sol)
 * [MagicInterface.sol](./MagicInterface.sol)
```

File: assets/eip-6732/contracts/Migrations.sol
[the remark this answers is not in the document]
Comment (SamWilsn) [reply assets/eip-6732/contracts/Migrations.sol:1]: That sounds reasonable, yes.

File: EIPS/eip-6732.md [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+     */
+    function isExpired(uint256 id) external view returns (bool);
+}
+```
+
+The **Supply extension** is OPTIONAL for compliant smart contracts. This allows contracts to associate a unique max supply for each token ID. Smart contracts implementing the `RedeemableTokenSupply` extension MUST abide by the following:
+
+- Implementers SHOULD NOT allow tokens to be minted if total supply exceeds max supply.
+- Implementers SHOULD increment total supply upon minting and decrement upon burning.
+- Implementers are RECOMMENDED to override the `_beforeMint` hook to increment total supply upon minting and decrement upon burning.
```
Comment (SamWilsn) [inline EIPS/eip-6732.md:188]: The `_beforeMint` function is part of the reference implementation, not the specification itself. The Specification section should be understandable on its own without the other sections, while the reference implementation should help to clarify what's in the the Specification section.

```suggestion
```

**Review by SamWilsn — CHANGES_REQUESTED** [review]
You'll also still need to remove external links and build files from your reference implementation.

**Comment by SamWilsn** [comment]
I am closing this pull request because we are in the process of separating EIPs and ERCs into distinct repositories. Unfortunately, as far as we are aware, GitHub does not provide any tools to ease this migration, so every pull request will need to be re-opened manually.

As this is a PR to create / modify an ERC, I will kindly ask you to redirect this to the new repository at [ethereum/ERCs](https://github.com/ethereum/ERCs). **We have prepared [a guide](https://github.com/ethereum/ERCs/issues/8) to help with the process.**

If there is relevant history here, please link to this PR from the new pull request.

On behalf of the EIP Editors, I apologize for this inconvenience.

**Pull request closed without merging**